# Nonparametric ML Models - Cumulative Lab

## Introduction

In this cumulative lab, you will apply two nonparametric models you have just learned — k-nearest neighbors and decision trees — to the forest cover dataset.

## Objectives

* Practice identifying and applying appropriate preprocessing steps
* Perform an iterative modeling process, starting from a baseline model
* Explore multiple model algorithms, and tune their hyperparameters
* Practice choosing a final model across multiple model algorithms and evaluating its performance

## Your Task: Complete an End-to-End ML Process with Nonparametric Models on the Forest Cover Dataset

![line of pine trees](images/trees.jpg)

Photo by <a href="https://unsplash.com/@michaelbenz?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText">Michael Benz</a> on <a href="/s/photos/forest?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText">Unsplash</a>

### Business and Data Understanding

To repeat the previous description:

> Here we will be using an adapted version of the forest cover dataset from the [UCI Machine Learning Repository](https://archive.ics.uci.edu/ml/datasets/covertype). Each record represents a 30 x 30 meter cell of land within Roosevelt National Forest in northern Colorado, which has been labeled as `Cover_Type` 1 for "Cottonwood/Willow" and `Cover_Type` 0 for "Ponderosa Pine". (The original dataset contained 7 cover types but we have simplified it.)

The task is to predict the `Cover_Type` based on the available cartographic variables:

In [1]:
# Run this cell without changes
import pandas as pd

df = pd.read_csv('/content/forest_cover.csv')
df

Elevation  Aspect  Slope  Horizontal_Distance_To_Hydrology  \
0           2553     235     17                               351   
1           2011     344     17                               313   
2           2022      24     13                               391   
3           2038      50     17                               408   
4           2018     341     27                               351   
...          ...     ...    ...                               ...   
38496       2396     153     20                                85   
38497       2391     152     19                                67   
38498       2386     159     17                                60   
38499       2384     170     15                                60   
38500       2383     165     13                                60   

       Vertical_Distance_To_Hydrology  Horizontal_Distance_To_Roadways  \
0                                  95                              780   
1                                  29                              404   
2                                  42                              509   
3                                  71                              474   
4                                  34                              390   
...                               ...                              ...   
38496                              17                              108   
38497                              12                               95   
38498                               7                               90   
38499                               5                               90   
38500                               4                               67   

       Hillshade_9am  Hillshade_Noon  Hillshade_3pm  \
0                188             253            199   
1                183             211            164   
2                212             212            134   
3                226             200            102   
4                152             188            168   
...              ...             ...            ...   
38496            240             237            118   
38497            240             237            119   
38498            236             241            130   
38499            230             245            143   
38500            231             244            141   

       Horizontal_Distance_To_Fire_Points  ...  Soil_Type_31  Soil_Type_32  \
0                                    1410  ...             0             0   
1                                     300  ...             0             0   
2                                     421  ...             0             0   
3                                     283  ...             0             0   
4                                     190  ...             0             0   
...                                   ...  ...           ...           ...   
38496                                 837  ...             0             0   
38497                                 845  ...             0             0   
38498                                 854  ...             0             0   
38499                                 864  ...             0             0   
38500                                 875  ...             0             0   

       Soil_Type_33  Soil_Type_34  Soil_Type_35  Soil_Type_36  Soil_Type_37  \
0                 0             0             0             0             0   
1                 0             0             0             0             0   
2                 0             0             0             0             0   
3                 0             0             0             0             0   
4                 0             0             0             0             0   
...             ...           ...           ...           ...           ...   
38496             0             0             0             0             0   
38497             0             0             0            

> As you can see, we have over 38,000 rows, each with 52 feature columns and 1 target column:

> * `Elevation`: Elevation in meters
> * `Aspect`: Aspect in degrees azimuth
> * `Slope`: Slope in degrees
> * `Horizontal_Distance_To_Hydrology`: Horizontal dist to nearest surface water features in meters
> * `Vertical_Distance_To_Hydrology`: Vertical dist to nearest surface water features in meters
> * `Horizontal_Distance_To_Roadways`: Horizontal dist to nearest roadway in meters
> * `Hillshade_9am`: Hillshade index at 9am, summer solstice
> * `Hillshade_Noon`: Hillshade index at noon, summer solstice
> * `Hillshade_3pm`: Hillshade index at 3pm, summer solstice
> * `Horizontal_Distance_To_Fire_Points`: Horizontal dist to nearest wildfire ignition points, meters
> * `Wilderness_Area_x`: Wilderness area designation (3 columns)
> * `Soil_Type_x`: Soil Type designation (39 columns)
> * `Cover_Type`: 1 for cottonwood/willow, 0 for ponderosa pine

This is also an imbalanced dataset, since cottonwood/willow trees are relatively rare in this forest:

In [2]:
# Run this cell without changes
print("Raw Counts")
print(df["Cover_Type"].value_counts())
print()
print("Percentages")
print(df["Cover_Type"].value_counts(normalize=True))

Raw Counts
0    35754
1     2747
Name: Cover_Type, dtype: int64

Percentages
0    0.928651
1    0.071349
Name: Cover_Type, dtype: float64


Thus, a baseline model that always chose the majority class would have an accuracy of over 92%. Therefore we will want to report additional metrics at the end.

### Previous Best Model

In a previous lab, we used SMOTE to create additional synthetic data, then tuned the hyperparameters of a logistic regression model to get the following final model metrics:

* **Log loss:** 0.13031294393913376
* **Accuracy:** 0.9456679825472678
* **Precision:** 0.6659919028340081
* **Recall:** 0.47889374090247455

In this lab, you will try to beat those scores using more-complex, nonparametric models.

### Modeling

Although you may be aware of some additional model algorithms available from scikit-learn, for this lab you will be focusing on two of them: k-nearest neighbors and decision trees. Here are some reminders about these models:

#### kNN - [documentation here](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html)

This algorithm — unlike linear models or tree-based models — does not emphasize learning the relationship between the features and the target. Instead, for a given test record, it finds the most similar records in the training set and returns an average of their target values.

* **Training speed:** Fast. In theory it's just saving the training data for later, although the scikit-learn implementation has some additional logic "under the hood" to make prediction faster.
* **Prediction speed:** Very slow. The model has to look at every record in the training set to find the k closest to the new record.
* **Requires scaling:** Yes. The algorithm to find the nearest records is distance-based, so it matters that distances are all on the same scale.
* **Key hyperparameters:** `n_neighbors` (how many nearest neighbors to find; too few neighbors leads to overfitting, too many leads to underfitting), `p` and `metric` (what kind of distance to use in defining "nearest" neighbors)

#### Decision Trees - [documentation here](https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html)

Similar to linear models (and unlike kNN), this algorithm emphasizes learning the relationship between the features and the target. However, unlike a linear model that tries to find linear relationships between each of the features and the target, decision trees look for ways to split the data based on features to decrease the entropy of the target in each split.

* **Training speed:** Slow. The model is considering splits based on as many as all of the available features, and it can split on the same feature multiple times. This requires exponential computational time that increases based on the number of columns as well as the number of rows.
* **Prediction speed:** Medium fast. Producing a prediction with a decision tree means applying several conditional statements, which is slower than something like logistic regression but faster than kNN.
* **Requires scaling:** No. This model is not distance-based. You also can use a `LabelEncoder` rather than `OneHotEncoder` for categorical data, since this algorithm doesn't necessarily assume that the distance between `1` and `2` is the same as the distance between `2` and `3`.
* **Key hyperparameters:** Many features relating to "pruning" the tree. By default they are set so the tree can overfit, and by setting them higher or lower (depending on the hyperparameter) you can reduce overfitting, but too much will lead to underfitting. These are: `max_depth`, `min_samples_split`, `min_samples_leaf`, `min_weight_fraction_leaf`, `max_features`, `max_leaf_nodes`, and `min_impurity_decrease`. You can also try changing the `criterion` to "entropy" or the `splitter` to "random" if you want to change the splitting logic.

### Requirements

#### 1. Prepare the Data for Modeling

#### 2. Build a Baseline kNN Model

#### 3. Build Iterative Models to Find the Best kNN Model

#### 4. Build a Baseline Decision Tree Model

#### 5. Build Iterative Models to Find the Best Decision Tree Model

#### 6. Choose and Evaluate an Overall Best Model

## 1. Prepare the Data for Modeling

The target is `Cover_Type`. In the cell below, split `df` into `X` and `y`, then perform a train-test split with `random_state=42` and `stratify=y` to create variables with the standard `X_train`, `X_test`, `y_train`, `y_test` names.

Include the relevant imports as you go.

In [3]:
# Your code here
X= df.drop('Cover_Type', axis=1)
y= df['Cover_Type']
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test= train_test_split(X, y,random_state=42 , stratify=y)

Now, instantiate a `StandardScaler`, fit it on `X_train`, and create new variables `X_train_scaled` and `X_test_scaled` containing values transformed with the scaler.

In [4]:
# Your code here
from sklearn.preprocessing import StandardScaler
scaler= StandardScaler()
X_train_scaled= scaler.fit_transform(X_train)
X_test_scaled= scaler.transform(X_test)

The following code checks that everything is set up correctly:

In [5]:
# Run this cell without changes

# Checking that df was separated into correct X and y
assert type(X) == pd.DataFrame and X.shape == (38501, 52)
assert type(y) == pd.Series and y.shape == (38501,)

# Checking the train-test split
assert type(X_train) == pd.DataFrame and X_train.shape == (28875, 52)
assert type(X_test) == pd.DataFrame and X_test.shape == (9626, 52)
assert type(y_train) == pd.Series and y_train.shape == (28875,)
assert type(y_test) == pd.Series and y_test.shape == (9626,)

# Checking the scaling
assert X_train_scaled.shape == X_train.shape
assert round(X_train_scaled[0][0], 3) == -0.636
assert X_test_scaled.shape == X_test.shape
assert round(X_test_scaled[0][0], 3) == -1.370

## 2. Build a Baseline kNN Model

Build a scikit-learn kNN model with default hyperparameters. Then use `cross_val_score` with `scoring="neg_log_loss"` to find the mean log loss for this model (passing in `X_train_scaled` and `y_train` to `cross_val_score`). You'll need to find the mean of the cross-validated scores, and negate the value (either put a `-` at the beginning or multiply by `-1`) so that your answer is a log loss rather than a negative log loss.

Call the resulting score `knn_baseline_log_loss`.

Your code might take a minute or more to run.

In [6]:
# Replace None with appropriate code

# Relevant imports
from sklearn. neighbors import KNeighborsClassifier
from sklearn.model_selection import cross_val_score

# Creating the model
knn_baseline_model = KNeighborsClassifier()

# Perform cross-validation
knn_baseline_log_loss = -(cross_val_score(knn_baseline_model,X_train_scaled , y_train, scoring="neg_log_loss").mean())

knn_baseline_log_loss

0.12964546386734577

In [7]:
knn_baseline_model.get_params()

{'algorithm': 'auto',
 'leaf_size': 30,
 'metric': 'minkowski',
 'metric_params': None,
 'n_jobs': None,
 'n_neighbors': 5,
 'p': 2,
 'weights': 'uniform'}

Our vanilla / baseline model uses k=5,p=2 and metric=minkowski

Our best logistic regression model had a log loss of 0.13031294393913376

Is this model better? Compare it in terms of metrics and speed.

In [8]:
# Replace None with appropriate text
"""
The knn model gives a log loss of 0.1296 which is lower than that of the logistic regression model thus its the better model
but in terms if speed, it is a bit slower than the logistic model
"""

'\nThe knn model gives a log loss of 0.1296 which is lower than that of the logistic regression model thus its the better model\nbut in terms if speed, it is a bit slower than the logistic model\n'

## 3. Build Iterative Models to Find the Best kNN Model

Build and evaluate at least two more kNN models to find the best one. Explain why you are changing the hyperparameters you are changing as you go. These models will be *slow* to run, so be thinking about what you might try next as you run them.

In [9]:
# Your code here (add more cells as needed)
#Let's try fitting a knn model with different values for k and determine the best k value
def find_best_k(X_train, y_train, X_test, y_test, min_k=1, max_k=25):
    # Your code here
    best_k=5
    best_score=0.12964546386734577
    odd_numbers = []
    for number in range(1, 26):
        if number % 2 != 0:
            odd_numbers.append(number)
            
    for k in odd_numbers:
        knn= KNeighborsClassifier(n_neighbors=k)
        # Fit the classifier
        knn.fit(X_train,y_train)
        # Predict on the test set
        test_preds = knn.predict(X_test)
        logloss =-(cross_val_score(knn,X_train , y_train, scoring="neg_log_loss").mean())
        
        if logloss <best_score:
            best_score=logloss
            best_k=k
            
    return best_k, best_score


find_best_k(X_train_scaled, y_train, X_test_scaled, y_test)


(19, 0.0636809988068612)

The best k is 19 with a log loss of 0.064

In [10]:
# Your code here (add more cells as needed)
#Lets try and fit a model with a different metric 
knn= KNeighborsClassifier(metric='euclidean')

logloss =-(cross_val_score(knn,X_train_scaled , y_train, scoring="neg_log_loss").mean())
logloss


knn= KNeighborsClassifier(metric='manhattan')

logloss =-(cross_val_score(knn,X_train_scaled , y_train, scoring="neg_log_loss").mean())
logloss


0.1181008016543303

using euclidean metric instead of minkowski which had a logloss of 0.12964546386734577 gives a logloss of 0.12964546386734577. thus there is no difference .
using manhattan distance gave a log loss of 0.1181008016543303 which is better than using euclidean and minkowski.

In [11]:
# Your code here (add more cells as needed)
knn= KNeighborsClassifier(n_neighbors=19,metric='manhattan')

logloss =-(cross_val_score(knn,X_train_scaled , y_train, scoring="neg_log_loss").mean())
logloss

0.06157974248852195

using k=19 and metric = manhattan gives us the best log loss

## 4. Build a Baseline Decision Tree Model

Now that you have chosen your best kNN model, start investigating decision tree models. First, build and evaluate a baseline decision tree model, using default hyperparameters (with the exception of `random_state=42` for reproducibility).

(Use cross-validated log loss, just like with the previous models.)

In [12]:
# Your code here
from sklearn.tree import DecisionTreeClassifier
clf= DecisionTreeClassifier(random_state=42)
model=clf.fit(X_train, y_train)

logloss =-(cross_val_score(model,X_train , y_train, scoring="neg_log_loss").mean())
logloss

0.7352281158853683

Interpret this score. How does this compare to the log loss from our best logistic regression and best kNN models? Any guesses about why?

In [13]:
# Replace None with appropriate text
"""
this model performs worse than knn and logistic regression. maybe the model is under/over fitting since we haven't done anym
model tuning
"""

"\nthis model performs worse than knn and logistic regression. maybe the model is under/over fitting since we haven't done anym\nmodel tuning\n"

## 5. Build Iterative Models to Find the Best Decision Tree Model

Build and evaluate at least two more decision tree models to find the best one. Explain why you are changing the hyperparameters you are changing as you go.

In [14]:
clf.get_params()

{'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': None,
 'max_leaf_nodes': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'random_state': 42,
 'splitter': 'best'}

In [15]:
# Your code here (add more cells as needed)
#lets try changing the criterion to be entropy
clf= DecisionTreeClassifier(criterion='entropy',random_state=42)
model=clf.fit(X_train, y_train)

logloss =-(cross_val_score(model,X_train , y_train, scoring="neg_log_loss").mean())
logloss

0.682800983683016

In [16]:
# Your code here (add more cells as needed)
# entropy criterion reduced the log loss so let's try incresing the max_depth
clf= DecisionTreeClassifier(max_depth=2,criterion='entropy',random_state=42)
model=clf.fit(X_train, y_train)

logloss =-(cross_val_score(model,X_train , y_train, scoring="neg_log_loss").mean())
logloss

0.17110532803415182

clearly increasing the tree depth from 0 to 2 made a significant difference , lets determine the best tree depth

In [17]:
def find_best_depth(X_train, y_train, X_test, y_test):
    # Your code here
    best_depth=2
    best_score=0.17110532803415182
    depth= list(range(1,26))
    for d in depth:
      clf= DecisionTreeClassifier(max_depth=d,criterion='entropy',random_state=42)
      logloss =-(cross_val_score(clf,X_train , y_train, scoring="neg_log_loss").mean())
      if logloss <best_score:
        best_score=logloss
        best_depth=d
            
    return best_depth, best_score
find_best_depth(X_train, y_train, X_test, y_test)

(6, 0.12045767599037507)

In [18]:
# Your code here (add more cells as needed)
#lets tweak min samples split
clf= DecisionTreeClassifier(min_samples_leaf= 3, max_depth=6,criterion='entropy',random_state=42)
model=clf.fit(X_train, y_train)

logloss =-(cross_val_score(model,X_train , y_train, scoring="neg_log_loss").mean())
logloss

0.11684995664951317

In [19]:
#tweaking min samples leaf reduces the logloss
def find_best_samples_leaf(X_train, y_train, X_test, y_test):
  best_leaf=3
  best_score=0.11684995664951317
  leaf= list(range(1,26))
  for d in leaf:
    clf= DecisionTreeClassifier(min_samples_leaf= d,max_depth=6,criterion='entropy',random_state=42)
    logloss =-(cross_val_score(clf,X_train , y_train, scoring="neg_log_loss").mean())
    if logloss <best_score:
      best_score=logloss
      best_leaf=d
    return best_leaf, best_score
find_best_samples_leaf(X_train, y_train, X_test, y_test)

(3, 0.11684995664951317)

The final decision tree classifier model has a logloss of 0.1168 which is lower than the logistic model but higher than the knn model. thus the best model is the knn model

## 6. Choose and Evaluate an Overall Best Model

Which model had the best performance? What type of model was it?

Instantiate a variable `final_model` using your best model with the best hyperparameters.

In [20]:
# Replace None with appropriate code
final_model = KNeighborsClassifier(n_neighbors=19,metric='manhattan')

# Fit the model on the full training data
# (scaled or unscaled depending on the model)
final_model.fit(X_train_scaled , y_train)


KNeighborsClassifier(metric='manhattan', n_neighbors=19)

Now, evaluate the log loss, accuracy, precision, and recall. This code is mostly filled in for you, but you need to replace `None` with either `X_test` or `X_test_scaled` depending on the model you chose.

In [21]:
# Replace None with appropriate code
from sklearn.metrics import accuracy_score, precision_score, recall_score, log_loss

preds = final_model.predict(X_test_scaled)
probs = final_model.predict_proba(X_test_scaled)

print("log loss: ", log_loss(y_test, probs))
print("accuracy: ", accuracy_score(y_test, preds))
print("precision:", precision_score(y_test, preds))
print("recall:   ", recall_score(y_test, preds))

log loss:  0.06511403266242219
accuracy:  0.9779763141491793
precision: 0.9018612521150592
recall:    0.7758369723435226


Interpret your model performance. How would it perform on different kinds of tasks? How much better is it than a "dummy" model that always chooses the majority class, or the logistic regression described at the start of the lab?

In [22]:
# Replace None with appropriate text
"""
The final model has a better log loss ,better accuracy of 98%, better precision of 90% and better recall of 77%
compared to the logistic model
accuracy: the model accurately predicts a value 98% of the time
precision: out of the predicted class 1 values, the model predicted 90% of them correctly
recall: out of the actual class 1 values , the model correctly predicted 78% of them
"""

'\nThe final model has a better log loss ,better accuracy of 98%, better precision of 90% and better recall of 77%\ncompared to the logistic model\naccuracy: the model accurately predicts a value 98% of the time\nprecision: out of the predicted class 1 values, the model predicted 90% of them correctly\nrecall: out of the actual class 1 values , the model correctly predicted 78% of them\n'

## Conclusion

In this lab, you practiced the end-to-end machine learning process with multiple model algorithms, including tuning the hyperparameters for those different algorithms. You saw how nonparametric models can be more flexible than linear models, potentially leading to overfitting but also potentially reducing underfitting by being able to learn non-linear relationships between variables. You also likely saw how there can be a tradeoff between speed and performance, with good metrics correlating with slow speeds.